File: bot.js
const TelegramBot = require('node-telegram-bot-api');
const ytdl = require('ytdl-core');
const ffmpeg = require('fluent-ffmpeg');
const fs = require('fs');
const path = require('path');
const youtubeSearch = require('youtube-search-api');

const token = '7163308622:AAGSibGMbW6hl2eQVhhuJRiBXyHXsqh0vGk';
const bot = new TelegramBot(token, { polling: true });

bot.onText(/\/start/, (msg) => {
    bot.sendMessage(msg.chat.id, "Welcome to the Song Download Bot. Send me a song name or a YouTube link to download the audio.");
});

bot.on('message', async (msg) => {
    const chatId = msg.chat.id;
    const text = msg.text;

    if (text.startsWith('http')) {
        downloadAndSendAudio(chatId, text);
    } else {
        const searchResults = await youtubeSearch.GetListByKeyword(text, false);
        if (searchResults.items.length > 0) {
            const videoUrl = `https://www.youtube.com/watch?v=${searchResults.items[0].id}`;
            downloadAndSendAudio(chatId, videoUrl);
        } else {
            bot.sendMessage(chatId, "No results found for your search. Please try with a different song name.");
        }
    }
});

function downloadAndSendAudio(chatId, url) {
    bot.sendMessage(chatId, "Downloading audio... Please wait.");

    const audioStream = ytdl(url, { filter: 'audioonly' });
    const audioFile = path.resolve(__dirname, 'audio.mp3');

    ffmpeg(audioStream)
        .audioBitrate(128)
        .save(audioFile)
        .on('end', () => {
            bot.sendAudio(chatId, audioFile).then(() => {
                fs.unlinkSync(audioFile);
            });
        })
        .on('error', (error) => {
            console.error(error);
            bot.sendMessage(chatId, "Failed to download audio.");
        });
}

File: bot.py
import os
import logging
from telegram import Update
from telegram.ext import Updater, CommandHandler, MessageHandler, Filters, CallbackContext
from pytube import YouTube, Search
from moviepy.editor import AudioFileClip

# Enable logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)
logger = logging.getLogger(__name__)

# Replace with your Telegram bot token
TOKEN = '7163308622:AAGSibGMbW6hl2eQVhhuJRiBXyHXsqh0vGk'

# Command to start the bot
def start(update: Update, context: CallbackContext) -> None:
    update.message.reply_text('Welcome to the Song Download Bot. Send me a song name or a YouTube link to download the audio.')

# Function to download and send audio
def download_and_send_audio(update: Update, url: str):
    chat_id = update.message.chat_id
    update.message.reply_text("Downloading audio... Please wait.")
    
    try:
        yt = YouTube(url)
        audio_stream = yt.streams.filter(only_audio=True).first()
        audio_file = audio_stream.download(filename='audio.mp4')
        
        # Convert to MP3
        audio_clip = AudioFileClip(audio_file)
        audio_clip.write_audiofile('audio.mp3')
        audio_clip.close()
        
        # Send the audio file to the user
        with open('audio.mp3', 'rb') as audio:
            context.bot.send_audio(chat_id=chat_id, audio=audio)
        
        # Clean up files
        os.remove('audio.mp4')
        os.remove('audio.mp3')
        
    except Exception as e:
        logger.error(e)
        update.message.reply_text("Failed to download audio.")

# Handle incoming messages
def handle_message(update: Update, context: CallbackContext) -> None:
    text = update.message.text
    
    # Check if the message is a valid YouTube URL
    if 'youtube.com/watch?v=' in text or 'youtu.be/' in text:
        download_and_send_audio(update, text)
    else:
        # If the message is not a URL, assume it's a song name and search for it on YouTube
        try:
            search_results = Search(text).results
            if search_results:
                video_url = search_results[0].watch_url
                download_and_send_audio(update, video_url)
            else:
                update.message.reply_text("No results found for your search. Please try with a different song name.")
        except Exception as e:
            logger.error(e)
            update.message.reply_text("Failed to search for the song. Please try again later.")

def main() -> None:
    # Create the Updater and pass it your bot's token
    updater = Updater(TOKEN)
    
    # Get the dispatcher to register handlers
    dispatcher = updater.dispatcher
    
    # on different commands - answer in Telegram
    dispatcher.add_handler(CommandHandler("start", start))
    
    # on noncommand i.e. message - handle the message
    dispatcher.add_handler(MessageHandler(Filters.text & ~Filters.command, handle_message))
    
    # Start the Bot
    updater.start_polling()
    
    # Run the bot until you press Ctrl-C or the process receives SIGINT, SIGTERM or SIGABRT
    updater.idle()

if __name__ == '__main__':
    main()
